lib/internal/url.js

```diff
@@ -1453,18 +1455,29 @@ function pathToFileURL(filepath) {
     outURL.hostname = domainToASCII(hostname);
     outURL.pathname = encodePathChars(
       RegExpPrototypeSymbolReplace(backslashRegEx, StringPrototypeSlice(filepath, hostnameEndIndex), '/'));
-  } else {
-    let resolved = path.resolve(filepath);
-    // path.resolve strips trailing slashes so we must add them back
-    const filePathLast = StringPrototypeCharCodeAt(filepath,
-                                                   filepath.length - 1);
-    if ((filePathLast === CHAR_FORWARD_SLASH ||
-         (isWindows && filePathLast === CHAR_BACKWARD_SLASH)) &&
-        resolved[resolved.length - 1] !== path.sep)
-      resolved += '/';
-    outURL.pathname = encodePathChars(resolved);
-  }
-  return outURL;
+    return outURL;
+  }
+  let resolved = path.resolve(filepath);
+  // path.resolve strips trailing slashes so we must add them back
+  const filePathLast = StringPrototypeCharCodeAt(filepath,
+                                                 filepath.length - 1);
+  if ((filePathLast === CHAR_FORWARD_SLASH ||
+       (isWindows && filePathLast === CHAR_BACKWARD_SLASH)) &&
+      resolved[resolved.length - 1] !== path.sep)
+    resolved += '/';
+
+  // Call encodePathChars first to avoid encoding % again for ? and #.
+  resolved = encodePathChars(resolved);
+
+  // Question and hash character should be included in pathname.
+  // Therefore, encoding is required to eliminate parsing them in different states.
+  // This is done as an optimization to not creating a URL instance and
+  // later triggering pathname setter, which impacts performance
+  if (StringPrototypeIndexOf(resolved, '?') !== -1)
+    resolved = RegExpPrototypeSymbolReplace(questionRegex, resolved, '%3F');
+  if (StringPrototypeIndexOf(resolved, '#') !== -1)
+    resolved = RegExpPrototypeSymbolReplace(hashRegex, resolved, '%23');
```

Comment:
Why isn't this part of `encodePathChars`?

Reply:
Because previous usage of encodePathChars sets the output of the function to pathname setter. In pathname setter, every character is assumed to be a pathname character. But in this case we are calling URL constructor directly where /path?query input will parse both pathname and query part of URL. In order to not do that, we need to encode it prior to calling the constructor, hence, creating a fast path of avoiding certain edge cases.